### PR TITLE
[libc] Add some integration features into newhdrgen

### PR DESCRIPTION
**Pull request by frobtech**
This adds some command-line switches to the newhdrgen script
that facilitate more generalized build system integration.
These enable the use of the script in Fuchsia's GN build.

The --write-if-changed switch avoids touching the .h output file
if it's unchanged, which is beneficial to incremental rebuilds.

The --depfile switch writes out a make/ninja-compatible depfile
describing the inputs and outputs.  This is crucial for build
system integration when using the following new feature.

When the new --libc-dir switch is given, the YAML file argument
is optional.  Instead, the --h_def_file input file path must have
the --libc-dir path as a prefix and the yaml file is found in its
corresponding place in the libc source tree.

### CODE REVIEW

**Review by michaelrj-google — COMMENTED** [review]
Code changes overall LGTM, make sure to update the docs where necessary: https://github.com/llvm/llvm-project/blob/main/libc/docs/dev/header_generation.rst?plain=1

File: libc/newhdrgen/yaml_to_classes.py [hunk 1]
```diff
@@ -252,30 +273,62 @@ def main():
...
+            or not path.exists()
+            or path.read_text() != contents
+        ):
+            path.write_text(contents)
+
+    yaml_file = args.yaml_file
+    if args.h_def_file and not yaml_file:
+        libc_include_dir = args.libc_dir / "include"
+        libc_yaml_dir = args.libc_dir / "newhdrgen" / "yaml"
+        yaml_file = libc_yaml_dir / args.h_def_file.with_suffix("").with_suffix(
```
[the remark this answers is not in the document]
Comment (frobtech) [reply libc/newhdrgen/yaml_to_classes.py:288]: `.with_suffix("")` strips `.def` and then `.with_suffix(".yaml")` replaces the remaining `.h` with `.yaml`.

**Comment by frobtech** [comment]
> Code changes overall LGTM, make sure to update the docs where necessary: https://github.com/llvm/llvm-project/blob/main/libc/docs/dev/header_generation.rst?plain=1

That does not currently include any explanation of the file-related command line switches. It's not really documentation for the tool as a tool. It's more a few examples of things to try and what might happen.

File: libc/newhdrgen/yaml_to_classes.py [hunk 2]
```diff
@@ -252,30 +273,62 @@ def main():
...
     if not args.export_decls and args.h_def_file:
         with open(args.h_def_file, "r") as f:
             h_def_content = f.read()
         final_header_content = fill_public_api(header_str, h_def_content)
-        with open(output_file_path, "w") as f:
-            f.write(final_header_content)
     else:
-        with open(output_file_path, "w") as f:
-            f.write(header_str)
+        final_header_content = str
```
Comment (fabio-d) [inline libc/newhdrgen/yaml_to_classes.py:330]: `final_header_content = header_str`?

File: libc/newhdrgen/yaml_to_classes.py [hunk 3]
```diff
@@ -252,30 +273,62 @@ def main():
...
+        args.depfile.write(f"{output_file_path}: {args.h_def_file} {args.yaml_file}\n")
+        args.depfile.close()
+
     if args.add_function:
         add_function_to_yaml(yaml_file, args.add_function)
 
     header_class = GpuHeader if args.export_decls else HeaderFile
     header = load_yaml_file(args.yaml_file, header_class, args.entry_points)
 
+    if args.add_function:
```
Comment (fabio-d) [inline libc/newhdrgen/yaml_to_classes.py:316]: Is this repetition (see 5 lines above) on purpose?

**Review by fabio-d — CHANGES_REQUESTED** [review]
I think it'd be great if the two blocks that decide `yaml_file` and `output_file_path` had some comment explaining the various cases with examples, because the code alone was a bit hard to follow to me

**Review by michaelrj-google — APPROVED** [review]
Overall LGTM, but you need to fix the paths before landing.

**Comment by frobtech** [comment]
I'm going to abandon this and do some more basic refactoring to get towards a good CLI for both programmatic and human use. It will be in several smaller steps.
